### PR TITLE
134 push entire dom

### LINKED ISSUE #134: Push DOM tree to Database

### PR DESCRIPTION
**Pull request by mpsijm**
Will close #134 

Changes:
- Copied JQuery-UI's zIndex source into our repo
- Add `DOMRewriter` class, which can recursively add `data-octopeer-*` tags to DOM elements
- Let `ContentController` use `DOMRewriter` to rewrite the DOM on specific moments:
  - on load
  - when the (new) `MutationObserver` notices a mutation in the DOM tree

If there's anything wrong with anything, tell me! :D Else, please merge :)

### CODE REVIEW

**Comment by mdingjan** [comment]
The additional lag that this feature takes with him is really noticable. xD
But that's something we already expected, so that's fine. ;)

**Comment by mdingjan** [comment]
The status of the extension gets quite buggy after switching a few pull requests, which could have two causes in my opinion:
- The lag 
- A new bug is born because of the changes. :bug:

I think the last one is most likely the cause.

**Comment by Exclaminator** [comment]
Nice coding style, LGTM.

File: src/octopeer-github/test/content/DOMRewriterTest.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/**
+ * Created by Maarten on 12-06-2016.
```
Comment (mdingjan) [inline src/octopeer-github/test/content/DOMRewriterTest.ts]: Thanks for adding the tests! :D
The other changes are related to the private and unreachable functions of the controllers, so I'm fine with that. :)

File: src/octopeer-github/main/MainController.ts [hunk 2]
```diff
@@ -166,6 +180,11 @@ class MainController implements OptionsObserver {
         Status.set(status);
     }
 
+    /**
+     * Sends to a tab whether to hook or unhook to/from the DOM.
+     * @param tab The tab to send this message to.
+     * @param hookToDom true if the tab shold hook to DOM, false if the tab should unhook from DOM.
```
Comment (mdingjan) [inline src/octopeer-github/main/MainController.ts]: There is a typo here; shold -> should.

**Comment by mdingjan** [comment]
The bug has indeed been fixed; thanks! :D 
I've played around more and can't find any more bugs, nice! :)
The functionality is there and it has been tested well now, so good job! :D 

Make sure to fix the typo and you have my approval for merging this PR. :)

File: src/octopeer-github/content/DOMRewriter.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     * @param rule The rule with which a node is transformed into a property value that is added to the DOM
+     * @returns {DOMRewriter} this
+     */
+    public addRule(attrName: string, rule: (node: JQuery) => string|number) {
+        this.rules[attrName] = rule;
+        return this;
+    }
+
+    /**
+     * Adds all rules that were thought useful during the Database Meetings: x, y, width, height and (optional) z.
```
Comment (MathiasMeuleman) [inline src/octopeer-github/content/DOMRewriter.ts]: 'during the Database Meetings' won't really make sense to other people. Make it somewhat generic

**Comment by MathiasMeuleman** [comment]
There isn't any possibility of making this faster right? The only problem is that I'm suddenly getting 400 statuscodes while typing this message, maybe something to look into?

**Comment by mpsijm** [comment]
Nope, this part cannot be made faster. Mainly because the rewriting of the DOM _has_ to happen in the content script, because the page cannot be reconstructed in the background page. 
The 400 status codes are because the space characters cannot be pushed to the database. I'll poke Aaron about that :P

**Comment by MathiasMeuleman** [comment]
Thanks for the rewrite, this does make it somewhat faster :) I will merge this

**Pull request merged**